iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java

```diff
@@ -277,19 +272,42 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
       if (serializedSnapshotsToPut != null) {
         List<Snapshot> snapshotsToPut =
             SnapshotsUtil.parseSnapshots(fileIO, serializedSnapshotsToPut);
-        Pair<List<Snapshot>, List<Snapshot>> snapshotsDiff =
-            SnapshotsUtil.symmetricDifferenceSplit(snapshotsToPut, updatedMetadata.snapshots());
-        List<Snapshot> appendedSnapshots = snapshotsDiff.getFirst();
-        List<Snapshot> deletedSnapshots = snapshotsDiff.getSecond();
-        snapshotInspector.validateSnapshotsUpdate(
-            updatedMetadata, appendedSnapshots, deletedSnapshots);
         Map<String, SnapshotRef> snapshotRefs =
             serializedSnapshotRefs == null
                 ? new HashMap<>()
                 : SnapshotsUtil.parseSnapshotRefs(serializedSnapshotRefs);
-        updatedMetadata =
-            maybeAppendSnapshots(updatedMetadata, appendedSnapshots, snapshotRefs, true);
-        updatedMetadata = maybeDeleteSnapshots(updatedMetadata, deletedSnapshots);
+
+        TableMetadata.Builder builder = TableMetadata.buildFrom(updatedMetadata);
+
+        // 1. Identify which snapshots are new vs existing
+        Set<Long> existingIds =
+            updatedMetadata.snapshots().stream()
```

Comment:
Nit: existingIds and updatedMetadata don't go well together. Maybe we should rename updatedMetadata to preUpdateMetadata or preparedMetadata